File: zsystem_customize/public/js/purchase_order.js
frappe.ui.form.on("Purchase Order",{
    supplier(frm){
        if(!frm.doc.supplier){
            return
        }
        frappe.call({
            method:"frappe.client.get_list",
            args : {
                "doctype": "Purchase Order",
                filters:{
                    supplier:frm.doc.supplier,
                    docstatus:["!=",2]
                },
                fields: ["name", "shipping_address"],
                order_by: "creation desc",
                limit_page_length: 1
            },
            callback:function(r){
                if(r.message && r.message.length){
                    let po = r.message[0];
                    if(po.shipping_address){
                        frm.set_value("shipping_address",po.shipping_address)
                    }
                }
            }
        })
         (frm.doc.items || []).forEach(row => {
            fetch_last_purchase_rate(frm, row.doctype, row.name);
        });
    },
    before_submit: function(frm) {
    return new Promise((resolve, reject) => {
        frappe.call({
            method: "zsystem_customize.purchase_order.validate_stock_against_qty",
            args: {
                doctype: frm.doctype,
                name: frm.doc.name,
            },
            callback: function(r) {
                if (!r.message || !r.message.has_error) {
                    resolve();
                    return;
                }

                let d = new frappe.ui.Dialog({
                    title: "Stock Warning",
                    size: "large",
                    fields: [
                        {
                            fieldtype: "HTML",
                            fieldname: "stock_html",
                            options: r.message.html
                        },
                        {
                            fieldtype: "Check",
                            fieldname: "confirm_submit",
                            label: "I have reviewed the stock warning and want to continue.",
                            onchange: function() {
                                let checked = d.get_value("confirm_submit");

                                if (checked) {
                                    d.get_primary_btn().show();
                                } else {
                                    d.get_primary_btn().hide();
                                }
                            }
                        }
                    ],
                    primary_action_label: "Submit",
                    primary_action() {
                        d.hide();
                        resolve();
                    }
                });

                d.show();

                // Hide Submit button initially
                d.get_primary_btn().hide();
            }
        });
    });
}
})
// last purchase rate and date
frappe.ui.form.on("Purchase Order Item", {
    item_code(frm, cdt, cdn) {
        fetch_last_purchase_rate(frm, cdt, cdn);
    }
});
function fetch_last_purchase_rate(frm, cdt, cdn) {
    const row = locals[cdt][cdn];

    if (!frm.doc.supplier || !row.item_code) {
        frappe.model.set_value(cdt, cdn, "custom_last_purchase_price", "");
        frappe.model.set_value(cdt, cdn, "custom_last_purchase_date", "");
        return;
    }

    frappe.call({
        method: "zsystem_customize.purchase_order.get_last_purchase_detail",
        args: {
            item_code: row.item_code,
            supplier: frm.doc.supplier
        },
        callback: function(r) {
            if (r.message && Object.keys(r.message).length) {
                frappe.model.set_value(cdt, cdn, "custom_last_purchase_price", r.message.rate);
                frappe.model.set_value(cdt, cdn, "custom_last_purchase_date", r.message.posting_date);
            } else {
                frappe.model.set_value(cdt, cdn, "custom_last_purchase_price", "");
                frappe.model.set_value(cdt, cdn, "custom_last_purchase_date", "");
            }
        }
    });
}

File: zsystem_customize/purchase_order.py
import frappe
# get last purchase rate and date

import frappe

@frappe.whitelist()
def get_last_purchase_detail(item_code, supplier):
    if not supplier:
        return {}

    data = frappe.db.sql("""
        SELECT
            pii.rate,
            pi.posting_date
        FROM `tabPurchase Invoice Item` pii
        INNER JOIN `tabPurchase Invoice` pi
            ON pi.name = pii.parent
        WHERE
            pii.item_code = %(item_code)s
            AND pi.supplier = %(supplier)s
            AND pi.docstatus = 1
        ORDER BY
            pi.posting_date DESC,
            pi.creation DESC
        LIMIT 1
    """, {
        "item_code": item_code,
        "supplier": supplier
    }, as_dict=True)

    return data[0] if data else {}

# //Validate so againsst po rate
def validate_so_rate(doc,method):
    for item in doc.items:
        if item.sales_order and item.item_code:
            so_rate = frappe.db.get_value("Sales Order Item",{"parent":item.sales_order,"item_code":item.item_code},"rate")
            if so_rate and item.rate >so_rate:
                frappe.throw(f"Row {item.idx}: Purchase Order Rate ({item.rate}) cannot be greater than Sales Order Rate ({so_rate}).")


# update po/material field based on so against po

def update_so_material_status(doc, method):

    sales_orders = {d.sales_order for d in doc.items if d.sales_order}

    if doc.docstatus == 2:  # cancelled
        for so in sales_orders:
            frappe.db.set_value("Sales Order", so, "custom_pomaterial_status", "")
        return

    if doc.docstatus != 1:  # only act on submit
        return

    for so in sales_orders:
        frappe.db.set_value(
            "Sales Order", so, "custom_pomaterial_status", "Waiting for Material"
        )


def update_so_material_status_on_receipt(doc, method):
    sales_orders = set()

    # ---------------------------------------------------------
    # Purchase Receipt
    # ---------------------------------------------------------
    if doc.doctype == "Purchase Receipt":

        for item in doc.items:
            if not item.purchase_order:
                continue

            po_items = frappe.db.get_all(
                "Purchase Order Item",
                filters={
                    "parent": item.purchase_order,
                    "sales_order": ["is", "set"]
                },
                fields=["sales_order"]
            )

            for po_item in po_items:
                if po_item.sales_order:
                    sales_orders.add(po_item.sales_order)

        # Cancelled Purchase Receipt
        if doc.docstatus == 2:
            status = "Waiting for Material"
        else:
            status = "Material Receive"

    # ---------------------------------------------------------
    # Purchase Invoice
    # ---------------------------------------------------------
    elif doc.doctype == "Purchase Invoice":

        # Only consider Purchase Invoice if Update Stock is checked
        if not doc.update_stock:
            return

        for item in doc.items:
            if not item.purchase_order:
                continue

            po_items = frappe.db.get_all(
                "Purchase Order Item",
                filters={
                    "parent": item.purchase_order,
                    "sales_order": ["is", "set"]
                },
                fields=["sales_order"]
            )

            for po_item in po_items:
                if po_item.sales_order:
                    sales_orders.add(po_item.sales_order)

        # Cancelled Purchase Invoice
        if doc.docstatus == 2:
            status = "Waiting for Material"
        else:
            status = "Material Receive"

    else:
        return

    # ---------------------------------------------------------
    # Update Sales Order
    # ---------------------------------------------------------
    for sales_order in sales_orders:
        frappe.db.set_value(
            "Sales Order",
            sales_order,
            "custom_pomaterial_status",
            status
        )

# Direct po validate for if already have qty in stock
@frappe.whitelist()
def validate_stock_against_qty(doctype, name):
    doc = frappe.get_doc(doctype, name)
    error_rows = []

    for item in doc.items:
        if item.sales_order:
            return

        actual_qty = frappe.db.get_value(
            "Item",
            {"item_code": item.item_code},
            "custom_stock_qty"
        ) or 0

        if item.qty <= actual_qty:
            error_rows.append(f"""
                <tr>
                    <td>{item.item_code}</td>
                    <td>{item.warehouse}</td>
                    <td>{item.qty}</td>
                    <td>{actual_qty}</td>
                </tr>
            """)

    if error_rows:
        html = f"""
            <h3>Already Stock for the following Items!</h3>
            <table class="table table-bordered" style="width:100%;margin-top:10px;font-size:13px">
                <thead>
                    <tr>
                        <th>Item Code</th>
                        <th>Warehouse</th>
                        <th>Required Qty</th>
                        <th>Available Qty</th>
                    </tr>
                </thead>
                <tbody>
                    {''.join(error_rows)}
                </tbody>
            </table>
            <br>
        """

        return {
            "has_error": True,
            "html": html
        }

    return {"has_error": False}

# //validate contact person no
def validate_contact_no(doc,method=None):
    mobile_no = doc.custom_contact_no
    if not mobile_no or not mobile_no.strip().isdigit() or len(mobile_no.strip()) != 10:
        frappe.throw(("Please enter a valid 10-digit contact person number."),title=("Invalid contact person Number"))
    
